### ISSUE TITLE
how to run model using  LlamaCpp from Langchain with gpu 

### ISSUE THREAD

**Issue by Muhamad-Nady** [abetlen/llama-cpp-python#199]
I am running GPT4ALL with LlamaCpp class which imported from langchain.llms, how i could use the gpu to run my model. because it has a very poor performance on cpu
could any one help me telling which dependencies i need to install, which parameters for LlamaCpp need to be changed
or high level apu not support the gpu for now
I have install the package with  "pip install llama-cpp-python"
could anyone help me with that, thanks in advance.

**Comment by gjmulder** [comment]
@Muhamad-Nady have you reviewed how to enable CuBLAS or a BLAS implementation in the [README.md](https://github.com/abetlen/llama-cpp-python#installation-with-openblas--cublas--clblast)? This is not full GPU support but will speed up inference for longer (> 32 word) prompts.

**Comment by Muhamad-Nady** [comment]
@gjmulder, yes i see it but i don't know which one should use with linux, windows. if i have NVIDIA is the same like i have a AMD GPU. I need a little of declaration, how i could use that.  I install  cuBLAS,   but i still have Blas here =0 "AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 | "

**Comment by gjmulder** [comment]
Can you please provide the complete description of your GPU hardware make, model, OS version, python version, text output of your build and install, as requested in the issue template? Ideally, without screenshots, but copy/pasted text so it is easier to read.

**Comment by Muhamad-Nady** [comment]
Hi @gjmulder here is what you asked for
os: Window 11 Home Version 22h2
GPU: Nvidia RTX 3050 laptop gpu
Python version: 3.10.9 anconda
build text : (base) C:\Users\Ryan\Documents\MuhamadTest>python guiwithmodel_fais.py
llama.cpp: loading model from C:\Users\Ryan\Documents\MuhamadTest\ggjt-model.bin
llama_model_load_internal: format = ggjt v1 (latest)
llama_model_load_internal: n_vocab = 32001
llama_model_load_internal: n_ctx = 2056
llama_model_load_internal: n_embd = 4096
llama_model_load_internal: n_mult = 256
llama_model_load_internal: n_head = 32
llama_model_load_internal: n_layer = 32
llama_model_load_internal: n_rot = 128
llama_model_load_internal: ftype = 2 (mostly Q4_0)
llama_model_load_internal: n_ff = 11008
llama_model_load_internal: n_parts = 1
llama_model_load_internal: model size = 7B
llama_model_load_internal: ggml ctx size = 59.11 KB
llama_model_load_internal: mem required = 5809.33 MB (+ 1026.00 MB per state)
llama_init_from_file: kv self size = 1028.00 MB
AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 
could you help me with that

**Comment by larawehbe** [comment]
@gjmulder  Hello
Im facing the same issue on Linux
OS: Rocky Linux 9 ( linode vm  )

GPU : Quadro RTX 6000 24GB 

I compiled the llama.cpp according to the readme file and specifically this one : 


``` mkdir build
cd build
cmake .. -DLLAMA_CUBLAS=ON
cmake --build . --config Release
``` 

The output has showed that "cuda" detected and worked upon it


When i run ./main interactive mode from inside llama.cpp, it works on gpu

When I run LlamaCppEmbeddings from LangChain and the same model (7b quantized ), it doesnt work on gpu and takes around 4minutes to answer a question using the RetrievelQAChain.


Any help in this part?

Thanks

**Comment by gjmulder** [comment]
1. Can you show the full output of each build and run as per the issue template? Use < details > < /details> if it is a lot of text.
2. What are you `--n_gpu_layers` set to for both runs?
3. Do you get the same behaviour if you compile and run the version of `llama.cpp` that is pulled in via `llama-cpp-python`?

**Comment by larawehbe** [comment]
> 

thanks for your response
```
embeddings = LlamaCppEmbeddings(model_path=original_model_path,
                              n_ctx=2048,
                              n_gpu_layers=24,
                              n_threads=8,
                              n_batch=1000)
llm = LlamaCpp(
    model_path=original_model_path,
    n_ctx= 2048,
    verbose=True,
    use_mlock=True,
    n_gpu_layers=12,
    n_threads=4,
    n_batch=1000
)
```

I did this, and its now working so so fast!!
Not sure why when i increase n_gpu_layers it starts to get slower, so for llm 8 was the fastest after several trial and errors

Now in the following
```
qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=retriever)
```
When i choose chain_type as "map_reduce", it becomes super slow

Any guidance why?

Thank you in advance

**Comment by gjmulder** [comment]
Run:

`watch -d nvidia-smi`

To see a high level overview of what's going on on your GPU that refreshes every 2 seconds.

You can select and periodically log states using something like:

`nvidia-smi -l 1 --query-gpu=name,index,utilization.gpu,utilization.memory,memory.used,temperature.gpu,power.draw --format=csv`

`man nvidia-smi` for all the details of what each metric means.

**Comment by Muhamad-Nady** [comment]
@gjmulder  I have a question. Depends on what I define the number of GPU layers?
I have nvidia rtx3050

**Comment by larawehbe** [comment]
> @gjmulder I have a question. Depends on what I define the number of GPU layers? I have nvidia rtx3050

Same question. Following

**Comment by gjmulder** [comment]
I don't know anything about Langchain, but if you're having multiple python processes running concurrently on the GPU, you need to ensure they all get enough:

1. CPU to feed them (`n_threads`)
2. VRAM for each context (`n_ctx`)
3. VRAM for each set of layers of the models you want to run on the GPU (`n_gpu_layers`)
4. GPU threads that the two GPU processes aren't saturating the GPU cores (this is unlikely to happen as far as I've seen)

`nvidia-smi` will tell you a lot about how the GPU is being loaded. I also keep an eye on `atop` to ensure the CPU isn't the bottleneck.
